homeassistant/components/fan/xiaomi_miio.py

```diff
@@ -196,50 +437,115 @@ def device_state_attributes(self):
 
     @property
     def is_on(self):
-        """Return true if fan is on."""
+        """Return true if device is on."""
         return self._state
 
-    @asyncio.coroutine
-    def _try_command(self, mask_error, func, *args, **kwargs):
-        """Call an air purifier command handling error messages."""
+    @staticmethod
+    def _extract_value_from_attribute(state, attribute):
+        from enum import Enum
+
+        value = getattr(state, attribute)
+        if isinstance(value, Enum):
+            return value.value
+
+        return value
+
+    async def _try_command(self, mask_error, func, *args, **kwargs):
+        """Call a miio device command handling error messages."""
         from miio import DeviceException
         try:
-            result = yield from self.hass.async_add_job(
+            result = await self.hass.async_add_job(
                 partial(func, *args, **kwargs))
 
-            _LOGGER.debug("Response received from air purifier: %s", result)
+            _LOGGER.debug("Response received from miio device: %s", result)
 
             return result == SUCCESS
         except DeviceException as exc:
             _LOGGER.error(mask_error, exc)
+            self._available = False
             return False
 
-    @asyncio.coroutine
-    def async_turn_on(self: ToggleEntity, speed: str = None, **kwargs) -> None:
-        """Turn the fan on."""
+    async def async_turn_on(self, speed: str = None,
+                            **kwargs) -> None:
+        """Turn the device on."""
         if speed:
             # If operation mode was set the device must not be turned on.
-            result = yield from self.async_set_speed(speed)
+            result = await self.async_set_speed(speed)
         else:
-            result = yield from self._try_command(
-                "Turning the air purifier on failed.", self._air_purifier.on)
+            result = await self._try_command(
+                "Turning the miio device on failed.", self._device.on)
 
         if result:
             self._state = True
             self._skip_update = True
```

Comment:
Worth adding a comment wrt `_skip_update` here also?

Reply:
The `skip_update` behavior is commented at the `async_update` already. I don't want to introduce a second comment here.